(feat) add repository metadata to repo match search result items

erzhtor opened the pull request:
Part of https://github.com/sourcegraph/sourcegraph/issues/43961.

This PR:
- Adds repository metadata (aka `keyValuePairs`) to search result items `commit|path|repo|symbol|path` if `repository-metadata` feature flag is enabled

## Test plan
- `sg start`
- Enable `repository-metadata` feature flag
- Add some key value pairs through API (currently no UI for adding) GQL mutation `addRepoKeyValuePair` API
- Search and check search results

## Screenshots
| Description | Screenshot |
| --: | :-- |
| Repo match | <img width="1728" alt="image" src="https://user-images.githubusercontent.com/6717049/231539259-6563b666-0609-4a3c-a5b1-c0dd43ba27dd.png"> |
| Commit match | <img width="1728" alt="image" src="https://user-images.githubusercontent.com/6717049/231539568-c4727d4d-8704-402c-a9b7-9dfcbfadba3b.png"> |
| Path match | <img width="1728" alt="image" src="https://user-images.githubusercontent.com/6717049/231539336-e6bf05fd-2aeb-42f9-b7dc-133d0868d61e.png"> |
| Symbol match | <img width="1728" alt="image" src="https://user-images.githubusercontent.com/6717049/231539388-a5b3a486-a4f8-4a8a-9b5d-ce172f86bf51.png"> |
| Content match | <img width="1728" alt="image" src="https://user-images.githubusercontent.com/6717049/231539517-22a8ec40-9b3e-4fbd-9d9a-af6955d92389.png"> |

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

## App preview:

- [Web](https://sg-web-erzhtor-add-repo-metadata-to.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

camdencheek reviewed (commented):
Woot! Thanks for tackling this.

A couple of quick first impressions:
1) It seems a little strange that we are adding this to all result types since they are logically associated only with the repos, not a file or a commit. I would expect these to only be displayed on repository result types.
2) The added gap between the title line and the result card is a little visually jarring. Might be worth a design consultation here
3) What does this look like with a null value? I don't see any screenshots with that case.

sg-e2e-regression-test-bob commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.01% (+2.13 kb) | 0.02% (+2.13 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/erzhtor/add-repo-metadata-to-search-results/compare-report.html;tab=live_object) for a full comparison between the commits 566f19d and 3f06a9b or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

limitedmage commented:
This is a great addition! I agree with Camden's comments here. I recommend reconsidering the design to put these key/value pairs inside the white box and only for repo results.

![CleanShot 2023-04-12 at 10 53 29](https://user-images.githubusercontent.com/206864/231542788-a6856ceb-fe44-47c2-932c-f0d0fd90e0cc.png)


I also recommend showing them in the repo page (eg. https://sourcegraph.com/github.com/sourcegraph/sourcegraph), but this can be a separate PR.

erzhtor commented:
Thanks for a quick review, @camdencheek.

> 1. It seems a little strange that we are adding this to all result types since they are logically associated only with the repos, not a file or a commit. I would expect these to only be displayed on repository result types.

I intentionally added to all result types except `own` since they all associate with repos, but we can remove and leave only for repo match too. cc @rrhyne @ryphil would appreciate your feedback.

> 2. The added gap between the title line and the result card is a little visually jarring. Might be worth a design consultation here

Yep, already asked Rob Rhyne for help with designs.

> 3. What does this look like with a null value? I don't see any screenshots with that case.

Good point, I missed that values are nullable, will take a look at this scenario too.

kopancek commented on client/branded/src/search-ui/components/ResultContainer.tsx:
```diff
@@ -35,6 +36,25 @@ const accessibleResultType: Record<SearchMatch['type'], string> = {
     team: 'team',
 }
 
+const RepoKVPs: React.FunctionComponent<{ keyValuePairs?: Record<string, string>; className?: string }> = ({
```
Can we call it `RepoMetadata` instead? KVPs is really cryptic on a first glance.

erzhtor replied:
I was following the db table naming, but agree that metadata sounds better, thanks.

kopancek commented on client/branded/src/search-ui/components/ResultContainer.tsx:
```diff
@@ -35,6 +36,25 @@ const accessibleResultType: Record<SearchMatch['type'], string> = {
+        return null
+    }
+    return (
+        <div className={classNames(styles.repoKvps, className, 'd-flex align-items-center flex-wrap')}>
```
same as above, I think it might be better to name this `styles.repoMeta` or similar:
```suggestion
        <div className={classNames(styles.repoMeta, className, 'd-flex align-items-center flex-wrap')}>
```

kopancek replied:
Instead of using a div with ton of styling, it might be easier to just use the existing Badge component from wildcard

kopancek commented on client/branded/src/search-ui/components/ResultContainer.module.scss:
```diff
@@ -59,3 +59,23 @@
+        font-size: 0.75rem;
+        padding: 0.075rem 0.2rem 0.075rem 0.375rem;
+        border-radius: 3px 0 0 3px;
+        color: var(--body-bg);
+        background-color: var(--text-muted);
```
Why did you reverse the colors?

kopancek commented on client/branded/src/search-ui/components/ResultContainer.module.scss:
```diff
@@ -59,3 +59,23 @@
+    gap: 0.5rem;
+
+    &-key {
+        font-size: 0.75rem;
+        padding: 0.075rem 0.2rem 0.075rem 0.375rem;
```
We do have variables and utility classes for font-sizes as well as standard padding sizes. Let's use those instead?

erzhtor replied:
Not relevant given https://github.com/sourcegraph/sourcegraph/pull/50567#discussion_r1165186271.

kopancek commented on client/branded/src/search-ui/components/ResultContainer.module.scss:
```diff
@@ -59,3 +59,23 @@
+.repo-kvps {
+    gap: 0.5rem;
+
+    &-key {
+        font-size: 0.75rem;
+        padding: 0.075rem 0.2rem 0.075rem 0.375rem;
+        border-radius: 3px 0 0 3px;
+        color: var(--body-bg);
+        background-color: var(--text-muted);
+    }
+
+    &-value {
+        font-size: 0.75rem;
+        padding: 0.075rem 0.375rem 0.075rem 0.2rem;
+        border-radius: 0 3px 3px 0;
+        color: var(--body-color);
+        background-color: var(--secondary);
+    }
```
There seems to be a lot of duplicated styles here. It might be easier to assign these styles to the parent container. That way you avoid duplication as well as the border-radius shenanigans.

erzhtor replied:
The only duplicate is font-size, the rest of the styles are different.

kopancek commented on client/branded/src/search-ui/components/FileContentSearchResult.tsx:
```diff
@@ -256,6 +256,7 @@ export const FileContentSearchResult: React.FunctionComponent<React.PropsWithChi
             className={classNames(styles.copyButtonContainer, containerClassName)}
             resultClassName={resultContainerStyles.highlightResult}
             rankingDebug={result.debug}
+            keyValuePairs={result.keyValuePairs}
```
If we call the feature repository metadata, why do we call everything that actually implements it key value pairs? Does not make sense to me to be honest.

erzhtor replied:
I believe metadata is implemented as simple key-value pairs for more flexibility.

kopancek reviewed (commented):
Mostly naming and styling questions/suggestions

sashaostrikov reviewed (approved):
Nice! LGTM after your changes and fixes.
I guess you'll move it as Juliana proposed in the next PR?

erzhtor commented:
Thanks, everyone, for the reviews and feedback. So after thinking a bit, 
- I followed @camdencheek suggestion on showing repository metadata only on RepoMatch results. We can add to other search result types if needed in future
  - Also added handling of null value cases (see screenshot for `oss` badge)
- I've followed @limitedmage suggestion and moved badges to header section
   - I'll also follow up in a separate PR @limitedmage's suggestion to add these badges to repo page too
- I've also followed @kopancek suggestion and reused badges with little styling from our current wildcard. Metadata key is `info` variant badge and value is `secondary` variant badge

Here is how it looks now:

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/6717049/231720425-d57f7fc7-3b34-4a49-9aa7-405f8e59ec0d.png">
